### ISSUE TITLE
JSDoc based generated typescript types missing

### ISSUE THREAD

**Issue by omgoshjosh** [jimmywarting/native-file-system-adapter#24]
Hello and thanks for this library!

I've been looking closely at this library and `fs-browser-access` and have decided to roll with this one because it unifies the folder picker with folder drop API, is standards based and forward looking, and more "bare metal" per @tomayac.

However, we're a typescript app and we really want to get those types in our project to make this library fit in better (no `@ts-ignore`!).

I see that the type generation was introduced a while back and it seems to have gone missing at some point.

[Here's the PR that first introduced the typescript types](https://github.com/jimmywarting/native-file-system-adapter/pull/5).

As I understand it, the preference for the maintainer is plain JS and a buildless world, which is nice but also probably a little less mainstream at this point in time.

When looking at the [history of the rollup build](https://github.com/jimmywarting/native-file-system-adapter/commits/master/rollup.config.js), I noticed a change to the rollup config from `src/es6.js` to `[src/**/*.js]` in a commit for a "big update" and think maybe that had an impact. 

I wouldn't mind contributing to get these types into my project.

@PaulTime brought them in originally over a year ago, maybe he can provide some guidance?

I'm unfamiliar with rollup and also unfamiliar with testing, developing, and publishing npm packages.

For reference, [here's typescript's docs on using JSDoc to produce typescript types](https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html).

[I am also looking into this package](https://www.npmjs.com/package/@types/wicg-file-system-access).

**Comment by jimmywarting** [comment]
Hi @omgoshjosh 👋 

Glad you picked this one. 

> I see that the type generation was introduced a while back and it seems to have gone missing at some point.

I will see if i can get this types back into the project if they somehow got. Many times i have been able to get typing support without any .d.ts files from 3th party modules that only have jsdoc, but i will try to get this .d.ts file back in

> As I understand it, the preference for the maintainer is plain JS and a buildless world, which is nice but also probably a little less mainstream at this point in time.

Yea, this is my preferences, Building something with esm and typescript is a nightmare to me, this hole path resolver is a total mess at this point, Deno land require extension, Typescript for node demand extensionless, some put in `.js` even doe no .js file exist cuz everything is in typescript, it don't run natively in any env, and is only a hassle to me. 
I support typescript and i even use it myself. but jsdoc + buildless is my preferences

**Comment by omgoshjosh** [comment]
thanks for the response!

🙇 🙇 🙇

i think i will be able to roll with the types from the definitely typed package @types/wicg-file-system-access for the time being. it appears to have the types pretty up to date and after a quick look through, it appears it matches the source you've published here.

**Comment by jimmywarting** [comment]
cool, doe it seems to be somewhat small when looking at the source?

**Comment by omgoshjosh** [comment]
here's the definitely typed entry:
https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/wicg-file-system-access

and here's the meat and potatoes index.d.ts file:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/wicg-file-system-access/index.d.ts

**Comment by dlech** [comment]
It looks like this has been fixed, however due to #27, it isn't available through npm yet.

**Comment by milahu** [comment]
the `wicg-file-system-access` types are incomplete

these are better
https://github.com/GoogleChromeLabs/browser-fs-access/blob/main/index.d.ts

compiled to jsdoc with https://github.com/futurGH/ts-to-jsdoc

<details>

```js
/**
 * Opens file(s) from disk.
 * @template {boolean} M
 * @param {
 *   | [FirstFileOpenOptions<M>, ...CoreFileOptions[]]
 *   | FirstFileOpenOptions<M>
 * } [options]
 * @returns {M extends boolean ? Promise<FileWithHandle> : Promise<FileWithHandle[]>}
 */
export function fileOpen(options) {
    throw new Error("FIXME this is a function signature from a *.d.ts declaration file. please move this to the *.js implementation file");
}
/**
 * Saves a file to disk.
 * @returns {Promise<any>} Optional file handle to save in place.
 * @param {Blob | Promise<Blob> | Response} blobOrPromiseBlobOrResponse
 * To-be-saved `Blob` or `Response`
 * @param {[FirstFileSaveOptions, ...CoreFileOptions[]] | FirstFileSaveOptions} [options]
 * @param {FileSystemFileHandle | null} [existingHandle]
 * A potentially existing file handle for a file to save to. Defaults to
 * `null`.
 * @param {boolean | false} [throwIfExistingHandleNotGood]
 * Determines whether to throw (rather than open a new file save dialog)
 * when `existingHandle` is no longer good. Defaults to `false`.
 * @param {(handle: FileSystemFileHandle | null) => void | null} [filePickerShown]
 * A callback to be called when the file picker was shown (which only happens
 * when no `existingHandle` is provided). Defaults to `null`.
 */
export function fileSave(blobOrPromiseBlobOrResponse, options, existingHandle, throwIfExistingHandleNotGood, filePickerShown) {
    throw new Error("FIXME this is a function signature from a *.d.ts declaration file. please move this to the *.js implementation file");
}
/**
 * Whether the File System Access API is supported.
 * @type {boolean}
 */
export const supported = (() => {
    throw new Error("FIXME this is a variable signature from a *.d.ts declaration file. please move this to the *.js implementation file");
})();
/**
 * @param {HTMLImageElement} img
 * @returns {Promise<any>}
 */
export function imageToBlob(img) {
    throw new Error("FIXME this is a function signature from a *.d.ts declaration file. please move this to the *.js implementation file");
}

/**
 * @typedef {
 *   | 'desktop'
 *   | 'documents'
 *   | 'downloads'
 *   | 'music'
 *   | 'pictures'
 *   | 'videos'
 * } WellKnownDirectory
 */

/** @typedef {'read' | 'readwrite'} FileSystemPermissionMode */

/**
 * Properties shared by all `options` provided to file save and open operations
 * @typedef {Object} CoreFileOptions
 * @property {string[]} [extensions] Acceptable file extensions. Defaults to `[""]`.
 * @property {string} [description] Suggested file description. Defaults to `""`.
 * @property {string[]} [mimeTypes] Acceptable MIME types. Defaults to `[]`.
 */

/**
 * Properties shared by the _first_ `options` object provided to file save and
 * open operations (any additional options objects provided to those operations
 * cannot have these properties)
 * @typedef {Object} FirstCoreFileOptions
 * @property {WellKnownDirectory|FileSystemHandle} [startIn]
 * @property {string} [id] By specifying an ID, the user agent can remember different directories for different IDs.
 * @property {boolean|false} [excludeAcceptAllOption] 
 */

/**
 * The first `options` object passed to file save operations can also specify
 * a filename
 * @typedef {Object} FirstFileSaveOptions
 * @property {string} [fileName] Suggested file name. Defaults to `"Untitled"`.
 * @property {(
 *   resolve: () => void,
 *   rejectionHandler: () => void,
 *   anchor: HTMLAnchorElement
 * ) => () => void} [legacySetup] 
 * Configurable cleanup and `Promise` rejector usable with legacy API for
 * determining when (and reacting if) a user cancels the operation. The
 * method will be passed a reference to the internal `rejectionHandler` that
 * can, e.g., be attached to/removed from the window or called after a
 * timeout. The method should return a function that will be called when
 * either the user chooses to open a file or the `rejectionHandler` is
 * called. In the latter case, the returned function will also be passed a
 * reference to the `reject` callback for the `Promise` returned by
 * `fileOpen`, so that developers may reject the `Promise` when desired at
 * that time.
 * Example rejector:
 *
 * const file = await fileOpen({
 *   legacySetup: (rejectionHandler) => {
 *     const timeoutId = setTimeout(rejectionHandler, 10_000);
 *     return (reject) => {
 *       clearTimeout(timeoutId);
 *       if (reject) {
 *         reject('My error message here.');
 *       }
 *     };
 *   },
 * });
 *
 * ToDo: Remove this workaround once
 *   https://github.com/whatwg/html/issues/6376 is specified and supported.
 */

/**
 * The first `options` object passed to file open operations can specify
 * whether multiple files can be selected (the return type of the operation
 * will be updated appropriately) and a way of handling cleanup and rejection
 * for legacy open operations.
 * @template {boolean} M
 * @typedef {Object} FirstFileOpenOptions
 * @property {M} [multiple] Allow multiple files to be selected. Defaults to `false`.
 * @property {(
 *   resolve: (
 *     value: M extends false | undefined ? FileWithHandle : FileWithHandle[]
 *   ) => void,
 *   rejectionHandler: () => void,
 *   input: HTMLInputElement
 * ) => (reject?: (reason?: any) => void) => void} [legacySetup] 
 * Configurable cleanup and `Promise` rejector usable with legacy API for
 * determining when (and reacting if) a user cancels the operation. The
 * method will be passed a reference to the internal `rejectionHandler` that
 * can, e.g., be attached to/removed from the window or called after a
 * timeout. The method should return a function that will be called when
 * either the user chooses to open a file or the `rejectionHandler` is
 * called. In the latter case, the returned function will also be passed a
 * reference to the `reject` callback for the `Promise` returned by
 * `fileOpen`, so that developers may reject the `Promise` when desired at
 * that time.
 * Example rejector:
 *
 * const file = await fileOpen({
 *   legacySetup: (rejectionHandler) => {
 *     const timeoutId = setTimeout(rejectionHandler, 10_000);
 *     return (reject) => {
 *       clearTimeout(timeoutId);
 *       if (reject) {
 *         reject('My error message here.');
 *       }
 *     };
 *   },
 * });
 *
 * ToDo: Remove this workaround once
 *   https://github.com/whatwg/html/issues/6376 is specified and supported.
 */

/** @typedef {Object} FileWithHandle
 * @property {FileSystemFileHandle} [handle] 
 */

/** @typedef {Object} FileWithDirectoryAndFileHandle
 * @property {FileSystemDirectoryHandle} [directoryHandle]
 * @property {FileSystemFileHandle} [handle] 
 */

/** @typedef {Object} FileSystemHandlePermissionDescriptor
 * @property {'read'|'readwrite'} [mode] 
 */

/** @typedef {Object} FileSystemHandle
 * @property {'file'|'directory'} kind
 * @property {string} name
 * @property {(other:FileSystemHandle)=>Promise<boolean>} isSameEntry
 * @property {(descriptor?:FileSystemHandlePermissionDescriptor)=>Promise<PermissionState>} queryPermission
 * @property {(descriptor?:FileSystemHandlePermissionDescriptor)=>Promise<PermissionState>} requestPermission 
 */
```

</details>
